dbus-wrapper/swupdd-main.c

```diff
@@ -0,0 +1,985 @@
+	                            char const * const opts_str[],
+	                            char const * const opts_bool[],
+				    char const * const opts_int[],
+				    struct list **strlist,
+				    sd_bus_error *error)
+{
+	int r;
+
+	r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Failed to enter options container");
+		return r;
+	}
+
+	while ((r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv")) > 0) {
+                const char *argname;
+
+                r = sd_bus_message_read(m, "s", &argname);
+                if (r < 0) {
+			sd_bus_error_set_errnof(error, -r, "Can't read option name");
+                        return r;
+		}
+		if (is_in_array(argname, opts_str)) {
+			r = bus_message_read_option_string(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else if (is_in_array(argname, opts_bool)) {
+			r = bus_message_read_option_bool(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else if (is_in_array(argname, opts_int)) {
+			r = bus_message_read_option_int(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else {
+			r = sd_bus_message_skip(m, "v");
+			if (r < 0) {
+				sd_bus_error_set_errnof(error, -r, "Can't skip unwanted option value");
+				return r;
+			}
+		}
+
+                r = sd_bus_message_exit_container(m);
+                if (r < 0) {
+			sd_bus_error_set_errnof(error, -r, "Can't exit dict entry container");
+                        return r;
+		}
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't exit options container");
+		return r;
+	}
+
+	return 0;
+}
+
+static int on_name_owner_change(sd_bus_message *m, void *userdata, sd_bus_error *ret_error) {
+	sd_event *e = userdata;
+
+	assert(m);
+	assert(e);
+
+	sd_bus_close(sd_bus_message_get_bus(m));
+	sd_event_exit(e, 0);
+
+	return 1;
+}
+
+static int on_childs_output(sd_event_source *s, int fd, uint32_t revents, void *userdata)
+{
+	daemon_state_t *context = userdata;
+	int r = 0;
+	char buffer[PIPE_BUF + 1];
+	ssize_t count;
+
+	if ((revents & EPOLLHUP) == EPOLLHUP) {
+		goto finish;
+	}
+
+	while ((count = read(fd, buffer, PIPE_BUF)) < 0 && (errno == EINTR)) {}
+	if (count > 0) {
+		buffer[count] = '\0';
+		r = sd_bus_emit_signal(context->bus,
+				       "/org/O1/swupdd/Client",
+				       "org.O1.swupdd.Client",
+				       "ChildOutputReceived", "s", buffer);
+		if (r < 0) {
+			ERR("Failed to emit signal: %s", strerror(-r));
+		}
+
+		return 0;
+	} else if (count < 0) {
+		ERR("Failed to read pipe: %s", strerror(errno));
+		/* Disable the handler since the pipe is broken */
+		r = -1;
+	}
+
+finish:
+	/* No more events for this handler are expected */
+	close(fd);
+	sd_event_source_unref(s);
+	return r;
+}
+
+static int on_child_exit(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata)
```

Comment:
This might be something you could use the https://www.freedesktop.org/software/systemd/man/sd_event_add_child.html instead of using a signal handler to avoid having to otherwise handle this case.